schema_salad/metaschema.py

```diff
@@ -587,28 +584,25 @@ def resolve(
         baseuri,  # type: str
         loadingOptions,  # type: LoadingOptions
     ):
-        # type: (...) -> Union[List[Union[Dict[str, str], str]], Dict[str, str], str]
-        m = self.typeDSLregex.match(doc)
-        if m:
-            group1 = m.group(1)
-            assert group1 is not None  # nosec
-            first = expand_url(group1, baseuri, loadingOptions, False, True, self.refScope)
-            second = third = None
-            if bool(m.group(2)):
-                second = {"type": "array", "items": first}
-                # second = CommentedMap((("type", "array"),
-                #                       ("items", first)))
-                # second.lc.add_kv_line_col("type", lc)
-                # second.lc.add_kv_line_col("items", lc)
-                # second.lc.filename = filename
-            if bool(m.group(3)):
-                third = ["null", second or first]
-                # third = CommentedSeq(["null", second or first])
-                # third.lc.add_kv_line_col(0, lc)
-                # third.lc.add_kv_line_col(1, lc)
-                # third.lc.filename = filename
-            return third or second or first
-        return doc
+        # type: (...) -> Union[List[Union[Dict[str, Any], str]], Dict[str, Any], str]
+        doc_ = doc
+        optional = False
+        if doc_.endswith("?"):
+            optional = True
+            doc_ = doc_[0:-1]
+
+        if doc_.endswith("[]"):
+            items = self.resolve(doc_[0:-2], baseuri, loadingOptions)
+            if isinstance(items, str):
+                items = expand_url(items, baseuri, loadingOptions, False, True, self.refScope)
+            expanded = {"type": "array", "items": items}  # type: Union[Dict[str, Any], str]
+        else:
+            expanded = expand_url(doc_, baseuri, loadingOptions, False, True, self.refScope)
+
+        if optional:
+            return ["null", expanded]
+        else:
+            return expanded
```

Comment:
Needs a switch for the old behavior if the schema-salad version is older than `1.3`

Reply:
How can I get the schema version?
I guess we need the `schemaVersion` field.

Reply:
Do you have any idea to implement `schemaVersion` field?

Ideally, we can introduce `schemaVersion` field at the top level as CWL spec does.
For the implicit context, it is easily implemented it by adding a schema for it.

The problem is when introducing it to the explicit context.
The explicit context accepts the `$base`, `$namespaces`, `$schemas`, and `$graph` fields and others are ignored.

In the current spec of SALAD, in my understanding, there are no way to introduce extra fields to the explicit context by the (meta) schema.
On the other hand, once we hard-code to accept the `schemaVersion` field in the explicit context, the packed CWL documents accidentally accept the `schemaVersion` field.

---
One workaround is just to introduce a new field `schemaVersion` to `LoadingOptions` and to add a new command line option such as `--schema-version` to `schema-salad-tool`.
At least we can implement it with a small effort and can provide a new feature to users soon.

Reply:
Oh, there already exists [`saladVersion`](https://github.com/common-workflow-language/cwl-v1.2/blob/c9d4ade6dd3725197b81ba30046fdcbae9f8002a/CommonWorkflowLanguage.yml#L1) but not documented!

We have to add a description of that but which repository to be fixed?
At least it has introduced since CWL v1.1.

Anyway, I can fix the typeDSL part to support older behavior by using `saladVersion`.

Reply:
yes!  We introduced saladVersion to handle future changes like this, but haven't used it.  In fact I don't think the schema salad codebase is even aware of its existence, but the CWL spec does contain saladVersion, so we can build that out.

Reply:
Uh...

If the generated SALAD parser must support SALAD v1.0-v1.3 and must change the behavior of typeDSL according to the `saladVersion` field, the problem I mentioned above still remains.

- The generated SALAD parser has to decide whether nested TypeDSL in a given SALAD file must be expanded or not.
  - It can be done by `saladVersion` field in the given SALAD file but there are no way to make it always enabled or disabled with the metaschema.
- Unfortunately the same limitation is in the generated CWL parser. That is, it decides whether nested TypeDSL must be expanded or not with `saladVersion`!
  - As explained above, there are no way to make the expansion of nested TypeDSL always enabled or disabled with the schema of CWL.

Is my understanding correct or are there any workaround for this?

Reply:
Done!